### PR TITLE
feat: update Evaluation struct to conform to encodable. Update logger.

### PR DESCRIPTION
**Pull request by PatrykPiwowarczyk**

### CODE REVIEW

File: Sources/FeaturevisorSDK/Instance.swift [hunk 1]
```diff
@@ -370,7 +409,7 @@ public class FeaturevisorInstance {
                     enabled: stickyFeature.enabled,
                     sticky: stickyFeature)
 
-                  logger.debug("using sticky enabled", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+                  logger.debug("using sticky enabled", ["featureKey": evaluation])
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:412]: @PatrykPiwowarczyk, you should map **evaluation** to dictionary and pass it here. Now, under "featureKey" key you are passing the entire evaluation.

The goal is to have something like this:
`logger.debug("using sticky enabled", evaluation.asDictionary())`

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 2]
```diff
@@ -4,6 +4,21 @@ import XCTest
 @testable import FeaturevisorTypes
 
 class FeaturevisorInstanceTests: XCTestCase {
+    
+    func testEncodeDoesNotThrow() throws {
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:8]: The name of this test it too generic. Based on this name we don't know what we encode. Remove "throws" too, as this is a thing which you want to assert

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 3]
```diff
@@ -4,6 +4,21 @@ import XCTest
...
+                    reason: .allocated,
+                    bucketValue: 42,
+                    ruleKey: "rule456",
+                    enabled: true,
+                    variableKey: "color"
+                )
+                
+        let encoder = JSONEncoder()
+        
+        XCTAssertNoThrow(try encoder.encode(evaluation))
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:20]: Not throwing an error doesn't mean that the encoding was done correctly. Can we make more detailed test?

Comment (PatrykPiwowarczyk) [reply Tests/FeaturevisorSDKTests/InstanceTests.swift:20]: I wanted to but evaluation would have to also conform to decodable which I didn't want to do only to test it. Unless there are some other ways

File: Sources/FeaturevisorSDK/Instance+Dictionary.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+//  
+//
+//  Created by Patryk Piwowarczyk on 28/09/2023.
+//
+
+import Foundation
+
+extension FeaturevisorInstance {
+
+    func toDictionary(_ evaluation: Evaluation) -> [String: Any] {
```
Comment (Rolfrider) [inline Sources/FeaturevisorSDK/Instance+Dictionary.swift:12]: I see you don't use this param

File: Sources/FeaturevisorSDK/Instance.swift [hunk 5]
```diff
@@ -672,7 +756,9 @@ public class FeaturevisorInstance {
                             bucketValue: bucketValue,
                             variation: variation)
 
-                        logger.debug("allocated variation", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+                    let evaluationDictionary = toDictionary(evaluation)
+
+                        logger.debug("allocated variation", ["featureKey": evaluationDictionary])
```
Comment (Rolfrider) [inline Sources/FeaturevisorSDK/Instance.swift:761]: too much tabs? 😜

**Review by Rolfrider — APPROVED** [review]

File: Sources/FeaturevisorSDK/Instance.swift [hunk 6]
```diff
@@ -81,6 +99,46 @@ public struct Evaluation {
...
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        featureKey = try container.decode(FeatureKey.self, forKey: .featureKey)
+        reason = try EvaluationReason(rawValue: container.decode(String.self, forKey: .reason)) ?? .error
+        bucketValue = try container.decodeIfPresent(BucketValue.self, forKey: .bucketValue)
+        ruleKey = try container.decodeIfPresent(RuleKey.self, forKey: .ruleKey)
+        error = try container.decodeIfPresent(String.self, forKey: .error).flatMap { NSError(domain: "", code: 0, userInfo: [NSLocalizedDescriptionKey: $0]) }
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:131]: Error is optional, so leave it as optional and don't create "dummy" error for it. It's confusing.

File: Sources/FeaturevisorSDK/Instance+Dictionary.swift (hunk 4 above)
Comment (polok) [inline Sources/FeaturevisorSDK/Instance+Dictionary.swift:12]: Such extension should be a part of Evaluation not FeaturevisorInstance

File: Sources/FeaturevisorSDK/Instance.swift [hunk 7]
```diff
@@ -370,7 +428,9 @@ public class FeaturevisorInstance {
                     enabled: stickyFeature.enabled,
                     sticky: stickyFeature)
 
-                  logger.debug("using sticky enabled", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+            let evaluationDictionary = toDictionary(evaluation)
+
+            logger.debug("using sticky enabled", ["featureKey": evaluationDictionary])
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:433]: You still assign everything to featureKey. As I mentioned earlier, it should be just like
`logger.debug("using sticky enabled",  evaluationDictionary`

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 8]
```diff
@@ -4,6 +4,63 @@ import XCTest
 @testable import FeaturevisorTypes
 
 class FeaturevisorInstanceTests: XCTestCase {
+    
+    func testEncodeEvaluationReturnsValidDatafileContent() throws {
+
+        let traffic = Traffic(key: "key", segments: .plain("segment"), percentage: 13, allocation: [])
+
+        let evaluation = Evaluation(
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:12]: IF we want to make a proper test, we should test more fields encoding

Comment (polok) [reply Tests/FeaturevisorSDKTests/InstanceTests.swift:12]: Also, please follow pattern which we have in other tests: GIVEN, WHEN, THEN

File: Sources/FeaturevisorSDK/Evaluation+Dictionary.swift [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+//  
+//
+//  Created by Patryk Piwowarczyk on 28/09/2023.
+//
+
+import Foundation
+
+extension Evaluation {
+
+    func toDictionary(_ evaluation: Evaluation) -> [String: Any] {
```
Comment (polok) [inline Sources/FeaturevisorSDK/Evaluation+Dictionary.swift:12]: This is an extension to Evaluation so you don't need to pass the same object as a `parameter.`

Also, do this in that way:
```
  func toDictionary() -> [String: Any] {
    guard let data = try? JSONEncoder().encode(self) else { return [:] }
    return (try? JSONSerialization.jsonObject(with: data, options: .allowFragments)).flatMap { $0 as? [String: Any] ?? [:]}
  }
```

File: Sources/FeaturevisorSDK/Instance.swift [hunk 10]
```diff
@@ -28,7 +28,24 @@ public enum EvaluationReason: String {
     case error
 }
 
-public struct Evaluation {
+public struct Evaluation: Codable {
+    private enum CodingKeys: String, CodingKey {
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:32]: Keep this at the bottom of Evaluation struct, after init(from decoder: Decoder)

File: Sources/FeaturevisorSDK/Instance.swift [hunk 11]
```diff
@@ -81,6 +99,46 @@ public struct Evaluation {
...
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        try container.encode(featureKey, forKey: .featureKey)
+        try container.encode(reason.rawValue, forKey: .reason)
+        try container.encode(bucketValue, forKey: .bucketValue)
+        try container.encode(ruleKey, forKey: .ruleKey)
+        if let error = error {
+            try container.encode(error.localizedDescription, forKey: .error)
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:111]: Why not: `try? container.encode(error.localizedDescription, forKey: .error)`?

Comment (polok) [reply Sources/FeaturevisorSDK/Instance.swift:111]: Actually, try? should be for all optional ones

File: Sources/FeaturevisorSDK/Instance.swift [hunk 12]
```diff
@@ -81,6 +99,46 @@ public struct Evaluation {
...
+        try container.encode(variationValue, forKey: .variationValue)
+        try container.encode(variableKey, forKey: .variableKey)
+        try container.encode(variableValue, forKey: .variableValue)
+        try container.encode(variableSchema, forKey: .variableSchema)
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        featureKey = try container.decode(FeatureKey.self, forKey: .featureKey)
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:127]: There are optional fields, so just `try` will throw an error if the field is missing. For optional ones, you should use `try?`

File: Sources/FeaturevisorSDK/Instance.swift [hunk 13]
```diff
@@ -370,7 +428,9 @@ public class FeaturevisorInstance {
                     enabled: stickyFeature.enabled,
                     sticky: stickyFeature)
 
-                  logger.debug("using sticky enabled", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+            let evaluationDictionary = evaluation.toDictionary(evaluation)
+
+            logger.debug("using sticky enabled", evaluationDictionary)
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:433]: Skip creating a variable as this is only for logging and invoke this as:
`logger.debug("using sticky enabled", evaluation.toDictionary())`

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 14]
```diff
@@ -5,6 +5,89 @@ import XCTest
...
+            variationValue: "value",
+            variableKey: "color",
+            variableValue: .string(""),
+            variableSchema: variableSchema
+        )
+
+        let encoder = JSONEncoder()
+        let jsonData = try encoder.encode(evaluation)
+
+        guard let decodedDictionary = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] else {
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:35]: This is your "when". Under "when" you put the code which you actually want to assert/test. So don't put there mock expected data (btw. it should be under "GIVEN" section) but this part

File: Sources/FeaturevisorSDK/Instance.swift [hunk 15]
```diff
@@ -51,25 +66,24 @@ public struct Evaluation {
...
+        enabled: Bool? = nil,
+        traffic: Traffic? = nil,
+        sticky: OverrideFeature? = nil,
+        initial: OverrideFeature? = nil,
+        variation: Variation? = nil,
+        variationValue: VariationValue? = nil,
+        variableKey: VariableKey? = nil,
+        variableValue: VariableValue? = nil,
+        variableSchema: VariableSchema? = nil
+    ) {
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:82]: Please, keep current style formating so move `) {` up

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 16]
```diff
@@ -5,6 +5,126 @@ import XCTest
...
+        //GIVEN
+        let evaluation = Evaluation(
+            featureKey: "feature123",
+            reason: .allocated
+        )
+
+        let expectedDictionary: [String: Any] = [
+            "featureKey": "feature123",
+            "reason": "allocated",
+            "bucketValue": NSNull(),
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:19]: Why not "nil"?

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 17]
```diff
@@ -5,6 +5,126 @@ import XCTest
...
+
+        //THEN
+        XCTAssertEqual(decodedDictionary as NSDictionary, expectedDictionary as NSDictionary)
+
+    }
+
+    func testEncodeEvaluationReturnsValidDatafileContent() throws {
+
+        //GIVEN
+        let traffic = Traffic(key: "key", segments: .plain("segment"), percentage: 13, allocation: [])
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:49]: As for the above, please format to:
```
Traffic(
    key: "key",
    segments: .plain("segment"),
    percentage: 13,
     allocation: [])
```
Easier to read and follow current code style formating

File: Sources/FeaturevisorSDK/Instance.swift [hunk 18]
```diff
@@ -80,6 +94,47 @@ public struct Evaluation {
         self.variableValue = variableValue
         self.variableSchema = variableSchema
     }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        try container.encode(featureKey, forKey: .featureKey)
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:101]: All fields which are optional should be encoded using `encodeIfPresent`

File: Tests/FeaturevisorSDKTests/InstanceTests.swift (hunk 16 above)
Comment (PatrykPiwowarczyk) [reply Tests/FeaturevisorSDKTests/InstanceTests.swift:19]: 'nil' is not compatible with expected dictionary value type 'Any'

File: Sources/FeaturevisorSDK/Instance.swift [hunk 19]
```diff
@@ -217,7 +280,70 @@ public class FeaturevisorInstance {
     }
 
     func getRevision() -> String {
-        return datafileReader.getRevision()
+        return self.datafileReader.getRevision()
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:283]: Why do we need `self` here?

File: Sources/FeaturevisorSDK/Instance.swift [hunk 20]
```diff
@@ -217,7 +280,70 @@ public class FeaturevisorInstance {
...
+    }
+
+    // MARK: - Bucketing
+
+    private func getFeature(byKey featureKey: String) -> Feature? {
```
Comment (polok) [inline Sources/FeaturevisorSDK/Instance.swift:288]: Whydo we have changes here and for `getBucketKey`?

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 21]
```diff
@@ -5,6 +5,137 @@ import XCTest
...
 
+    func testEncodingEvaluationWithNilValuesReturnsValidDatafileContent() throws {
+
+        //GIVEN
+        let evaluation = Evaluation(
+            featureKey: "feature123",
+            reason: .allocated
+        )
+
+        let expectedDictionary: [String: Any] = [
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:16]: Put expected under // Then. Thanks in advance.

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 22]
```diff
@@ -5,6 +5,137 @@ import XCTest
...
+            sticky: overrideFeature,
+            initial: overrideFeature,
+            variation: variation,
+            variationValue: "value",
+            variableKey: "color",
+            variableValue: .string(""),
+            variableSchema: variableSchema
+        )
+
+        let mockedTraffic: [String: Any] = [
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:79]: Don't use mocked as it's confusing. Rather `expectedTraffic`. Also move expected stuff under "THEN"

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 23]
```diff
@@ -5,6 +5,137 @@ import XCTest
...
+
+        //WHEN
+        guard
+            let decodedDictionary = try JSONSerialization.jsonObject(with: jsonData, options: [])
+                as? [String: Any]
+        else {
+            XCTFail("Failed to decode JSON to a dictionary.")
+            return
+        }
+
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:136]: You don't have "THEN" where you can assert vs `expectedDictionary` as now it's not being used

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 24]
```diff
@@ -14,11 +145,17 @@ class FeaturevisorInstanceTests: XCTestCase {
         // WHEN
         XCTAssertThrowsError(try createInstance(options: options)) { error in
 
-            // THEN
-            XCTAssertEqual(
-                error as! FeaturevisorError,
-                FeaturevisorError.invalidURL(string: "hf://wrong url.com")
-            )
+            MockURLProtocol.requestHandler = { request in
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:148]: Why such changes here?

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 25]
```diff
@@ -52,23 +188,6 @@ class FeaturevisorInstanceTests: XCTestCase {
             return (response, jsonString.data(using: .utf8))
         }
 
-        var featurevisorOptions = InstanceOptions.default
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:55]: Why we have here these changes?

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 26]
```diff
@@ -828,8 +947,18 @@ class FeaturevisorInstanceTests: XCTestCase {
             Thread.sleep(forTimeInterval: 0.1)
         }
 
-        // THEN
-        XCTAssertEqual(refreshedCount, expectedRefreshCount)
+        MockURLProtocol.requestHandler = { request in
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:950]: Why we have here these changes?

File: Tests/FeaturevisorSDKTests/InstanceTests.swift [hunk 27]
```diff
@@ -877,28 +1006,6 @@ class FeaturevisorInstanceTests: XCTestCase {
         XCTAssertEqual(isRefreshingStopped, true)
     }
 
-    func testSetDatafileByValidJSON() {
```
Comment (polok) [inline Tests/FeaturevisorSDKTests/InstanceTests.swift:880]: Why we have these changes here?

**Pull request closed without merging**